rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbCreateRelationNode.java

```diff
@@ -71,167 +84,56 @@ protected boolean createEntityIfNotExists() {
+            var createRelationFuture = Futures.transformAsync(deleteCurrentRelationsIfNeeded(ctx, msg.getOriginator(), relationType), __ ->
+                    checkRelationAndCreateIfAbsent(ctx, originator, targetEntityId, relationType), MoreExecutors.directExecutor());
+            return Futures.transform(createRelationFuture, createdOrAlreadyExists -> {
+                if (createdOrAlreadyExists && config.isChangeOriginatorToRelatedEntity()) {
+                    TbMsg tbMsg = ctx.transformMsgOriginator(msg, targetEntityId);
+                    return new TbPair<>(tbMsg, true);
                 }
-            }
-            return Futures.transform(Futures.allAsList(list), result -> null, ctx.getDbCallbackExecutor());
-        }, ctx.getDbCallbackExecutor());
-    }
-
-    private ListenableFuture<Boolean> checkRelationAndCreateIfAbsent(TbContext ctx, EntityContainer entityContainer, String relationType, SearchDirectionIds sdId) {
-        return Futures.transformAsync(checkRelation(ctx, sdId, relationType), relationPresent -> {
-            if (relationPresent) {
-                return Futures.immediateFuture(true);
-            }
-            return processCreateRelation(ctx, entityContainer, sdId, relationType);
-        }, ctx.getDbCallbackExecutor());
+                return new TbPair<>(msg, createdOrAlreadyExists);
+            }, MoreExecutors.directExecutor());
+        }, MoreExecutors.directExecutor());
     }
 
-    private ListenableFuture<Boolean> checkRelation(TbContext ctx, SearchDirectionIds sdId, String relationType) {
-        return ctx.getRelationService().checkRelationAsync(ctx.getTenantId(), sdId.getFromId(), sdId.getToId(), relationType, RelationTypeGroup.COMMON);
-    }
-
-    private ListenableFuture<Boolean> processCreateRelation(TbContext ctx, EntityContainer entityContainer, SearchDirectionIds sdId, String relationType) {
-        switch (entityContainer.getEntityType()) {
-            case ASSET:
-                return processAsset(ctx, entityContainer, sdId, relationType);
-            case DEVICE:
-                return processDevice(ctx, entityContainer, sdId, relationType);
-            case CUSTOMER:
-                return processCustomer(ctx, entityContainer, sdId, relationType);
-            case DASHBOARD:
-                return processDashboard(ctx, entityContainer, sdId, relationType);
-            case ENTITY_VIEW:
-                return processView(ctx, entityContainer, sdId, relationType);
-            case EDGE:
-                return processEdge(ctx, entityContainer, sdId, relationType);
-            case TENANT:
-                return processTenant(ctx, entityContainer, sdId, relationType);
-            case USER:
-                return processUser(ctx, entityContainer, sdId, relationType);
+    private ListenableFuture<Void> deleteCurrentRelationsIfNeeded(TbContext ctx, EntityId originator, String relationType) {
```

Comment:
move logic that is common for delete and create relation node to abstract class.